larq/optimizers_v2.py

```diff
@@ -1,10 +1,121 @@
+import warnings
+
 import tensorflow as tf
 import larq as lq
 
 from larq import utils
 from copy import deepcopy
 
 
+@utils.register_keras_custom_object
+class CaseOptimizer(tf.keras.optimizers.Optimizer):
+    """A case operation for optimizers that each train a subset of a model's variables.
+
+    An optimizer is used to train a variable iff its accompanying predicate evaluates to
+    `True`.
+
+    For each variable, at most one optimizer's predicate may evaluate to `True`. If no
+    optimizer's predicate evaluates to `True` for a variable, it is trained with the
+    `default` optimizer.
+
+    # Arguments
+    pred_opt_pairs: a list of `(tf.keras.optimizers.Optimzer, pred)` pairs, where `pred` 
+        takes one `tf.Variable` as argument and returns `True` if the optimizer should
+        train that variable, e.g. `pred(var) === True`.
+    default: a `tf.keras.optimizers.Optimizer` to be applied to any variable not claimed
+        by any other optimizer.
+    """
+
+    def __init__(self, pred_opt_pairs, default, name="optim_case"):
+        super().__init__(name=name)
+
+        for i, (pred, opt) in enumerate(pred_opt_pairs):
+            if not callable(pred):
+                raise TypeError(
+                    f"Expected callable predicate at `pred_opt_pairs[{i}][0]` but got `{type(pred)}`."
+                )
+            if not isinstance(opt, tf.keras.optimizers.Optimizer):
+                raise TypeError(
+                    f"Expected `tf.keras.optimizers.Optimizer` at `pred_opt_pairs[{i}][1]` but got `{type(opt)}`."
+                )
+        self.pred_opt_pairs = pred_opt_pairs
+
+        if default is not None and not isinstance(
+            default, tf.keras.optimizers.Optimizer
+        ):
+            raise TypeError(
+                f"Expected `tf.keras.optimizers.Optimizer` for `default` but got `{type(fallback_optimizer)}`."
+            )
+        self.default = default
+
+    def __getattr__(self, name):
+        if name == "lr":
+            raise ValueError("No single learning rate available for CaseOptimizer.")
+        return super().__getattr__(name)
+
+    def apply_gradients(self, grads_and_vars, name=None):
+        opt_grads_and_vars = [[] for _ in range(len(self.pred_opt_pairs))]
+        default_grads_and_vars = []
+
+        for grad, var in grads_and_vars:
+            num_opts = 0
+            for i, (pred, opt) in enumerate(self.pred_opt_pairs):
+                if pred(var):
+                    opt_grads_and_vars[i].append((grad, var))
+                    num_opts += 1
+
+            if num_opts == 0:
+                default_grads_and_vars.append((grad, var))
```

Comment:
I think we should throw here if no default optimizer is defined.

Reply:
We have this a few lines later:

```python
        if len(default_grads_and_vars) > 0 and self.default is None:
            warnings.warn(
                f"No `default` provided to train variables `{default_grads_and_vars}`."
            )
```

@koenhelwegen mentioned that not having a default could be desired behavior, e.g. for the usecase of only retraining batch norms and not touching other weights. That's why we changed the error throw I had here originally to a warning.

Reply:
:+1: Make sense, we can keep it as is.